commit 09bb98ff417c01f6250c2103431ca43170e71471
Author: Hongli Lai (Phusion) <hongli@phusion.nl>
Date:   Tue May 21 21:24:11 2013 +0200

    Fix some native packaging issues.
    The Ruby extension's source code is now included in the fakeroot.
    Introduced a 'libdir' locations.ini option.
    The Phusion Passenger object files are now included in the fakeroot.

diff --git a/bin/passenger-config b/bin/passenger-config
index e42196f..1821fe1 100755
--- a/bin/passenger-config
+++ b/bin/passenger-config
@@ -46,14 +46,14 @@ def common_library
 	require 'phusion_passenger/common_library'
 	return COMMON_LIBRARY.
 		only(*NGINX_LIBS_SELECTOR).
-		set_output_dir("#{PhusionPassenger.source_root}/libout/common/libpassenger_common")
+		set_output_dir("#{PhusionPassenger.lib_dir}/common/libpassenger_common")
 end
 
 case ARGV[0]
 when "--root"
 	puts PhusionPassenger.source_root
 when "--nginx-libs"
-	text = "#{common_library.link_objects_as_string} #{PhusionPassenger.source_root}/libout/common/libboost_oxt.a"
+	text = "#{common_library.link_objects_as_string} #{PhusionPassenger.lib_dir}/common/libboost_oxt.a"
 	if PhusionPassenger::PlatformInfo.has_math_library?
 		text << " -lm"
 	end
@@ -64,7 +64,7 @@ when "--compiled"
 			exit 1
 		end
 	end
-	if File.exist?("#{PhusionPassenger.source_root}/libout/common/libboost_oxt.a")
+	if File.exist?("#{PhusionPassenger.lib_dir}/common/libboost_oxt.a")
 		exit 0
 	else
 		exit 1
diff --git a/build/packaging.rb b/build/packaging.rb
index ff5685e..1177e8e 100644
--- a/build/packaging.rb
+++ b/build/packaging.rb
@@ -181,23 +181,29 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	# We don't use CONFIG['archdir'] and the like because we want
 	# the files to be installed to /usr, and the Ruby interpreter
 	# on the packaging machine might be in /usr/local.
-	fake_libdir = "#{fakeroot}/usr/lib/ruby/#{CONFIG['ruby_version']}"
+	fake_rubylibdir = "#{fakeroot}/usr/lib/ruby/vendor_ruby"
+	fake_libdir = "#{fakeroot}/usr/lib/phusion-passenger"
 	fake_native_support_dir = "#{fakeroot}/usr/lib/ruby/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
-	fake_agents_dir = "#{fakeroot}#{NATIVELY_PACKAGED_AGENTS_DIR}"
-	fake_helper_scripts_dir = "#{fakeroot}#{NATIVELY_PACKAGED_HELPER_SCRIPTS_DIR}"
+	fake_agents_dir = "#{fakeroot}/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
+	fake_helper_scripts_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
 	fake_resources_dir = "#{fakeroot}/usr/share/phusion-passenger"
-	fake_docdir = "#{fakeroot}#{NATIVELY_PACKAGED_DOC_DIR}"
+	fake_docdir = "#{fakeroot}/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}"
 	fake_bindir = "#{fakeroot}/usr/bin"
 	fake_sbindir = "#{fakeroot}/usr/sbin"
-	fake_apache2_module = "#{fakeroot}#{NATIVELY_PACKAGED_APACHE2_MODULE}"
-	fake_apache2_module_dir = File.dirname(fake_apache2_module)
+	fake_apache2_module_dir = "#{fakeroot}/usr/lib/apache2/modules"
+	fake_apache2_module = "#{fake_apache2_module_dir}/mod_passenger.so"
+	fake_ruby_extension_source_dir = "#{fakeroot}/usr/share/phusion-passenger/ruby_extension_source"
 	
 	sh "rm -rf #{fakeroot}"
 	sh "mkdir -p #{fakeroot}"
 	
+	sh "mkdir -p #{fake_rubylibdir}"
+	sh "cp #{PhusionPassenger.ruby_libdir}/phusion_passenger.rb #{fake_rubylibdir}/"
+	sh "cp -R #{PhusionPassenger.ruby_libdir}/phusion_passenger #{fake_rubylibdir}/"
+
 	sh "mkdir -p #{fake_libdir}"
-	sh "cp #{PhusionPassenger.ruby_libdir}/phusion_passenger.rb #{fake_libdir}/"
-	sh "cp -R #{PhusionPassenger.ruby_libdir}/phusion_passenger #{fake_libdir}/"
+	sh "cp -R #{PhusionPassenger.lib_dir}/common #{fake_libdir}/"
+	sh "rm -rf #{fake_libdir}/common/libboost_oxt"
 	
 	sh "mkdir -p #{fake_native_support_dir}"
 	native_support_archdir = PlatformInfo.ruby_extension_binary_compatibility_id
@@ -208,6 +214,7 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	sh "cp -R #{PhusionPassenger.agents_dir}/* #{fake_agents_dir}/"
 	sh "rm -rf #{fake_agents_dir}/*.dSYM"
 	sh "rm -rf #{fake_agents_dir}/*/*.dSYM"
+	sh "rm -rf #{fake_agents_dir}/*.o"
 	
 	sh "mkdir -p #{fake_helper_scripts_dir}"
 	sh "cp -R #{PhusionPassenger.helper_scripts_dir}/* #{fake_helper_scripts_dir}/"
@@ -234,6 +241,23 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	sh "mkdir -p #{fake_apache2_module_dir}"
 	sh "cp #{APACHE2_MODULE} #{fake_apache2_module_dir}/"
 
+	sh "mkdir -p #{fake_ruby_extension_source_dir}"
+	sh "cp -R #{PhusionPassenger.ruby_extension_source_dir}/* #{fake_ruby_extension_source_dir}"
+
+	puts "Creating #{fake_rubylibdir}/phusion_passenger/locations.ini"
+	File.open("#{fake_rubylibdir}/phusion_passenger/locations.ini", "w") do |f|
+		f.puts "natively_packaged=true"
+		f.puts "bin=/usr/bin"
+		f.puts "agents=/usr/lib/phusion-passenger/agents"
+		f.puts "libdir=/usr/lib/phusion-passenger"
+		f.puts "helper_scripts=/usr/share/phusion-passenger/helper-scripts"
+		f.puts "resources=/usr/share/phusion-passenger"
+		f.puts "doc=/usr/share/doc/phusion-passenger"
+		f.puts "rubylibdir=/usr/lib/ruby/vendor_ruby"
+		f.puts "apache2_module=/usr/lib/apache2/modules/mod_passenger.so"
+		f.puts "ruby_extension_source=/usr/share/phusion-passenger/ruby_extension_source"
+	end
+
 	sh "find #{fakeroot} -name .DS_Store -print0 | xargs -0 rm -f"
 end
 
diff --git a/doc/Packaging.txt.md b/doc/Packaging.txt.md
index 6a3ced2..e2e3fc9 100644
--- a/doc/Packaging.txt.md
+++ b/doc/Packaging.txt.md
@@ -84,13 +84,14 @@ The location configuration file is an ini file that looks as follows:
     [locations]
     natively_packaged=true
     bin=/usr/bin
-    agents=/usr/lib/phusion-passenger
+    agents=/usr/lib/phusion-passenger/agents
+    libdir=/usr/lib/phusion-passenger
     helper_scripts=/usr/share/phusion-passenger/helper-scripts
     resources=/usr/share/phusion-passenger
-    doc=/usr/share/doc
-    rubylibdir=/usr/lib/ruby/1.9.0
+    doc=/usr/share/doc/phusion-passenger
+    rubylibdir=/usr/lib/ruby/vendor_ruby
     apache2_module=/usr/lib/apache2/modules/mod_passenger.so
-    ruby_extension_source=/usr/share/phusion_passenger/ruby_native_support_source
+    ruby_extension_source=/usr/share/phusion-passenger/ruby_extension_source
 
 All keys except fo `natively_packaged` specify the locations of assets and asset
 directories. The "Asset types" section provides a description of all asset types.
@@ -192,6 +193,13 @@ a list of all possible assets and asset directories.
 
    Value when originally packaged: `<SOURCE_ROOT>/doc`.
 
+ * `libdir`
+
+   A directory that contains the Phusion Passenger library files, e.g.
+   libboost_oxt.a and various .o files.
+
+   Value when originally packaged: `<SOURCE_ROOT>/libout`
+
  * `rubylibdir`
 
    A directory that contains the Phusion Passenger Ruby library files. Note that
@@ -249,6 +257,3 @@ You can generate a fakeroot with the command `rake fakeroot`. This will
 generate an FHS-compliant directory tree in `pkg/fakeroot`, which you can
 directly package or with minor modifications. The fakeroot even contains
 a location configuration file.
-
-If the default fakeroot structure is not sufficient, please consider
-sending a patch.
diff --git a/lib/phusion_passenger.rb b/lib/phusion_passenger.rb
index 662aa2b..edf5ba0 100644
--- a/lib/phusion_passenger.rb
+++ b/lib/phusion_passenger.rb
@@ -62,15 +62,6 @@ module PhusionPassenger
 	# System-wide directory for storing Phusion Passenger Standalone runtime files.
 	GLOBAL_STANDALONE_RESOURCE_DIR = "/var/lib/#{GLOBAL_STANDALONE_NAMESPACE_DIRNAME}".freeze
 	
-	NATIVELY_PACKAGED_BIN_DIR                = "/usr/bin".freeze
-	NATIVELY_PACKAGED_AGENTS_DIR             = "/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents".freeze
-	NATIVELY_PACKAGED_HELPER_SCRIPTS_DIR     = "/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts".freeze
-	NATIVELY_PACKAGED_RESOURCES_DIR          = "/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}".freeze
-	NATIVELY_PACKAGED_DOC_DIR                = "/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}".freeze
-	NATIVELY_PACKAGED_RUNTIME_LIBDIR         = "/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}".freeze
-	NATIVELY_PACKAGED_HEADER_DIR             = "/usr/include/#{GLOBAL_NAMESPACE_DIRNAME}".freeze
-	NATIVELY_PACKAGED_APACHE2_MODULE         = "/usr/lib/apache2/modules/mod_passenger.so".freeze
-	
 	# Follows the logic of ext/common/ResourceLocator.h, so don't forget to modify that too.
 	def self.locate_directories(source_root_or_location_configuration_file = nil)
 		source_root_or_location_configuration_file ||= find_location_configuration_file
@@ -97,6 +88,7 @@ module PhusionPassenger
 			@natively_packaged     = get_bool_option(filename, options, 'natively_packaged')
 			@bin_dir               = get_option(filename, options, 'bin').freeze
 			@agents_dir            = get_option(filename, options, 'agents').freeze
+			@lib_dir               = get_option(filename, options, 'libdir').freeze
 			@helper_scripts_dir    = get_option(filename, options, 'helper_scripts').freeze
 			@resources_dir         = get_option(filename, options, 'resources').freeze
 			@doc_dir               = get_option(filename, options, 'doc').freeze
@@ -107,6 +99,7 @@ module PhusionPassenger
 			@natively_packaged     = false
 			@bin_dir               = "#{@source_root}/bin".freeze
 			@agents_dir            = "#{@source_root}/agents".freeze
+			@lib_dir               = "#{@source_root}/libout".freeze
 			@helper_scripts_dir    = "#{@source_root}/helper-scripts".freeze
 			@resources_dir         = "#{@source_root}/resources".freeze
 			@doc_dir               = "#{@source_root}/doc".freeze
@@ -138,6 +131,10 @@ module PhusionPassenger
 	def self.agents_dir
 		return @agents_dir
 	end
+
+	def self.lib_dir
+		return @lib_dir
+	end
 	
 	def self.helper_scripts_dir
 		return @helper_scripts_dir
